server/mailrepository/mailrepository-blob/src/main/scala/org/apache/james/mailrepository/blob/BlobMailRepository.scala

```diff
@@ -0,0 +1,186 @@
+        Mono.from(blobStoreDAO.save(bucketName, blobId, payload)).`then`(Mono.just(blobId))
+
+
+      LazyList(Pair.of(MailPartsId.METADATA_BLOB_TYPE, save)).asJavaSeqStream
+    }
+  }
+
+  private[blob] class MailDecoder private[blob](blobIdFactory: BlobId.Factory)
+    extends Impl.Decoder[(Mail, MimeMessagePartsId)] {
+
+    private def readMail(mailMetadata: MailMetadata): Mail = {
+      val builder = MailImpl.builder
+        .name(mailMetadata.name)
+        .sender(mailMetadata.sender.map(MaybeSender.getMailSender).getOrElse(MaybeSender.nullSender))
+        .addRecipients(mailMetadata.recipients.map(new MailAddress(_)).asJavaCollection)
+        .remoteAddr(mailMetadata.remoteAddr)
+        .remoteHost(mailMetadata.remoteHost)
+
+      mailMetadata.state.foreach(builder.state)
+      mailMetadata.errorMessage.foreach(builder.errorMessage)
+
+      mailMetadata.lastUpdated.map(Date.from).foreach(builder.lastUpdated)
+
+      mailMetadata.attributes.foreach { case (name, value) => builder.addAttribute(new Attribute(AttributeName.of(name), AttributeValue.fromJsonString(value))) }
+
+      builder.addAllHeadersForRecipients(retrievePerRecipientHeaders(mailMetadata.perRecipientHeaders))
+
+      builder.build
+    }
+
+
+    private def retrievePerRecipientHeaders(perRecipientHeaders: Map[String, Iterable[Header]]): PerRecipientHeaders = {
+      val result = new PerRecipientHeaders()
+      perRecipientHeaders.foreach { case (key, value) =>
+        value.foreach(headers => {
+          headers.values.foreach(header => {
+            val builder = PerRecipientHeaders.Header.builder().name(headers.key).value(header)
+            result.addHeaderForRecipient(builder, new MailAddress(key))
+          })
+        })
+      }
+      result
+    }
+
+
+    import serializers._
+
+    override def decode(streams: util.Map[BlobType, Store.CloseableByteSource]): (Mail, MimeMessagePartsId) = {
+      val source = streams.get(MailPartsId.METADATA_BLOB_TYPE)
+      val value = Json.fromJson[MailMetadata](Json.parse(source.openStream())).get
+      (readMail(value), value.mimePartsId(blobIdFactory))
+    }
+  }
+}
+
+class BlobMailRepository(val blobStore: BlobStoreDAO,
+                         val blobIdFactory: BlobId.Factory,
+                         val mimeMessageStore: Store[MimeMessage, MimeMessagePartsId]
+                        ) extends MailRepository {
+  private val bucketName: BucketName = BucketName.of("mailMetadata")
+  private val mimeMessageBucketName: BucketName = BucketName.of("mimeMessageData")
+  import BlobMailRepository._
+
+  @throws[MessagingException]
+  override def store(mc: Mail): MailKey = {
+    mimeMessageStore.save(mc.getMessage)
+      .flatMap(mimePartsId => mailMetadataStore.save((mc, mimePartsId)))
+      .map(mailPartsId => mailPartsId.toMailKey)
+      .block()
+  }
+
+  @throws[MessagingException]
+  override def size: Long = Flux.from(blobStore.listBlobs(bucketName)).count().block()
```

Comment:
There might be a higher level S3 method for getting bucket size, that might be more efficient? Would we benefit from adding it to the blobStore API ?

Reply:
I could not find one